zkevm-circuits/src/evm_circuit/execution/add.rs

```diff
@@ -0,0 +1,229 @@
+use crate::{
+    evm_circuit::{
+        execution::{
+            bus_mapping_tmp::{Block, Call, ExecStep, Transaction},
+            ExecutionGadget,
+        },
+        step::ExecutionState,
+        util::{
+            common_gadget::SameContextGadget,
+            constraint_builder::{
+                ConstraintBuilder, StepStateTransition, Transition::Delta,
+            },
+            math_gadget::{AddWordsGadget, PairSelectGadget},
+            select,
+        },
+    },
+    util::Expr,
+};
+use bus_mapping::evm::OpcodeId;
+use halo2::{arithmetic::FieldExt, circuit::Region, plonk::Error};
+
+// AddGadget verifies ADD and SUB at the same time by an extra swap flag,
+// when it's ADD, we annotate stack as [a, b, ...] and [c, ...],
+// when it's SUB, we annotate stack as [c, b, ...] and [a, ...].
+// Then we verify if a + b is equal to c.
+#[derive(Clone, Debug)]
+pub(crate) struct AddGadget<F> {
+    same_context: SameContextGadget<F>,
+    add_words: AddWordsGadget<F, 2>,
+    is_sub: PairSelectGadget<F>,
+}
+
+impl<F: FieldExt> ExecutionGadget<F> for AddGadget<F> {
+    const NAME: &'static str = "ADD";
+
+    const EXECUTION_STATE: ExecutionState = ExecutionState::ADD;
+
+    fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
+        let opcode = cb.query_cell();
+
+        let a = cb.query_word();
+        let b = cb.query_word();
+        let add_words = AddWordsGadget::construct(cb, [a.clone(), b.clone()]);
+        let c = add_words.sum();
+
+        // Swap a and c if opcode is SUB
+        let is_sub = PairSelectGadget::construct(
+            cb,
+            opcode.expr(),
+            OpcodeId::SUB.expr(),
+            OpcodeId::ADD.expr(),
+        );
+
+        // ADD: Pop a and b from the stack, push c on the stack
+        // SUB: Pop c and b from the stack, push a on the stack
+        cb.stack_pop(select::expr(is_sub.expr().0, c.expr(), a.expr()));
+        cb.stack_pop(b.expr());
+        cb.stack_push(select::expr(is_sub.expr().0, a.expr(), c.expr()));
+
+        // State transition
+        let step_state_transition = StepStateTransition {
+            rw_counter: Delta(3.expr()),
```

Comment:
Do we have a plan to handle these constant numbers in future PR?

Reply:
You mean a mapping from opcode to these constant like the constant gas? I think that will be helpful for stack only opcodes.

Reply:
Yes